contracts/partners/tokenWrappers/PufferPointTokenWrapper.sol

```diff
@@ -0,0 +1,106 @@
+/// impact would be to forfeit these tokens
+contract PufferPointTokenWrapper is BaseMerklTokenWrapper {
+    using SafeERC20 for IERC20;
+
+    // ================================= CONSTANTS =================================
+
+    mapping(address => VestingData) public vestingData;
+    uint256 public cliffDuration;
+    address public underlying;
+
+    // ================================= FUNCTIONS =================================
+
+    function initializeWrapper(address _underlying, uint256 _cliffDuration, IAccessControlManager _core) public {
+        super.initialize(_core);
+        if (_underlying == address(0)) revert ZeroAddress();
+        underlying = _underlying;
+        cliffDuration = _cliffDuration;
+    }
+
+    function token() public view override returns (address) {
+        return underlying;
+    }
+
+    function _beforeTokenTransfer(address from, address to, uint256 amount) internal override {
+        // Needs an underlying approval beforehand, this is how mints of wrappers are done
+        if (to == DISTRIBUTOR) {
+            IERC20(underlying).safeTransferFrom(from, address(this), amount);
+            _mint(from, amount); // These are then transferred to the distributor
+        }
+
+        // Will be burn right after, to avoid having any token aside from on the distributor
+        if (to == FEE_RECIPIENT) {
+            IERC20(underlying).safeTransferFrom(from, FEE_RECIPIENT, amount);
+            _mint(from, amount); // These are then transferred to the fee manager
+        }
+    }
+
+    function _afterTokenTransfer(address from, address to, uint256 amount) internal override {
+        if (to == FEE_RECIPIENT) {
+            _burn(to, amount); // To avoid having any token aside from on the distributor
+        }
+
+        if (from == DISTRIBUTOR) {
+            _burn(to, amount);
+            _createVesting(to, amount);
+        }
+    }
+
+    function _createVesting(address user, uint256 amount) internal {
+        VestingData storage userVestingData = vestingData[user];
+        VestingID[] storage userAllVestings = userVestingData.allVestings;
+        userAllVestings.push(VestingID(uint128(amount), uint128(block.timestamp + cliffDuration)));
+    }
+
+    function claim(address user) external {
+        VestingData storage userVestingData = vestingData[user];
+        VestingID[] storage userAllVestings = userVestingData.allVestings;
+        uint256 i = userVestingData.nextClaimIndex;
+        uint256 claimable;
+        while (true) {
+            VestingID storage userCurrentVesting = userAllVestings[i];
+            if (userCurrentVesting.unlockTimestamp > block.timestamp) {
+                claimable += userCurrentVesting.amount;
+                ++i;
+            } else {
+                userVestingData.nextClaimIndex = i;
+                break;
+            }
+        }
+        IERC20(token()).safeTransfer(user, claimable);
+    }
```

Comment:
Gas issue can lead to revert.
Would advise either capping the number of iteration or creating a new function to be able to claim a specific set of indexes.